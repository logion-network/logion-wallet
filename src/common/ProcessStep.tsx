import { ButtonVariant } from 'react-bootstrap/types';

import Dialog from './Dialog';
import { Action } from './Button';
import { Choice } from './MultiChoiceButton';
import { Children } from './types/Helpers';

export interface NextStep {
    id: string,
    callback?: () => void,
    mayProceed: boolean,
    buttonVariant: ButtonVariant,
    buttonText: string,
    choices?: Choice[],
}

export interface Props {
    active: boolean,
    title?: string,
    children: Children,
    nextSteps?: NextStep[],
    hasSideEffect?: boolean,
}

function toAction(nextStep: NextStep): Action {
    if(nextStep.callback === undefined && (nextStep.choices === undefined || nextStep.choices.length === 0)) {
        throw new Error("The step does nothing");
    }
    return {
        id: nextStep.id,
        callback: nextStep.callback,
        disabled: !nextStep.mayProceed,
        buttonVariant: nextStep.buttonVariant,
        buttonText: nextStep.buttonText,
        choices: nextStep.choices
    };
}

export default function ProcessStep(props: Props) {

    if(!props.active) {
        return null;
    }

    let nextSteps: NextStep[] = props.nextSteps || [];
    if(nextSteps.length === 0 && (props.hasSideEffect === undefined || !props.hasSideEffect)) {
        throw new Error("This process step is a trap");
    }

    return (
        <Dialog
            show={ props.active }
            size="lg"
            actions={ nextSteps.map(nextStep => toAction(nextStep)) }
        >
            <>
                { props.title && <h2>{ props.title }</h2> }
                { props.children }
            </>
        </Dialog>
    );
}
