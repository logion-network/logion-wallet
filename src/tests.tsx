import { ReactElement } from 'react';
import { createRenderer } from 'react-test-renderer/shallow';
import reactRenderer, { ReactTestRenderer, ReactTestRendererJSON, act as reactAct } from 'react-test-renderer';
import { screen, waitFor } from '@testing-library/react';
import userEvent from '@testing-library/user-event';

export function render(element: ReactElement): null | ReactTestRendererJSON | ReactTestRendererJSON[] {
    const component = reactRenderer.create(element);
    return component.toJSON();
}

export function renderer(element: ReactElement): ReactTestRenderer {
    return reactRenderer.create(element);
}

export function shallowRender<E extends ReactElement>(element: ReactElement): E {
    const renderer = createRenderer();
    renderer.render(element);
    return renderer.getRenderOutput();
}

export function mockAccount(address: string, name: string) {
    return {
        address,
        meta: {
            name
        }
    };
}

export { reactAct as act };

export async function clickByName(name: string | RegExp | ((accessibleName: string, element: Element) => boolean)) {
    let button: HTMLElement;
    await waitFor(() => button = screen.getByRole("button", { name }));
    await userEvent.click(button!);
}

export async function typeByLabel(label: string, value: string) {
    let input: HTMLElement;
    await waitFor(() => input = screen.getByLabelText(label));
    await userEvent.type(input!, value);
}
