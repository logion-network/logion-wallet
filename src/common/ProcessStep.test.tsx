import { shallowRender } from '../tests';

import ProcessStep from './ProcessStep';

test("Rendering inactive step", () => {
    const proceedCallback = jest.fn();
    const tree = shallowRender(
        <ProcessStep
            active={ false }
            title="Title"
            nextSteps={[
                {
                    id: "proceed",
                    buttonText: "Proceed",
                    buttonVariant: "primary",
                    mayProceed: true,
                    callback: proceedCallback
                }
            ]}
        >
            <p>Content</p>
        </ProcessStep>
    );
    expect(tree).toMatchSnapshot();
});

test("Rendering non-closable active step, may proceed", () => {
    const proceedCallback = jest.fn();
    const tree = shallowRender(
        <ProcessStep
            active={ true }
            title="Title"
            nextSteps={[
                {
                    id: "proceed",
                    buttonText: "Proceed",
                    buttonVariant: "primary",
                    mayProceed: true,
                    callback: proceedCallback
                }
            ]}
        >
            <p>Content</p>
        </ProcessStep>
    );
    expect(tree).toMatchSnapshot();
});

test("Rendering non-closable active step, may not proceed", () => {
    const proceedCallback = jest.fn();
    const tree = shallowRender(
        <ProcessStep
            active={ true }
            title="Title"
            nextSteps={[
                {
                    id: "proceed",
                    buttonText: "Proceed",
                    buttonVariant: "primary",
                    mayProceed: false,
                    callback: proceedCallback
                }
            ]}
        >
            <p>Content</p>
        </ProcessStep>
    );
    expect(tree).toMatchSnapshot();
});

test("Trap detection", () => {
    expect(() => shallowRender(
        <ProcessStep
            active={ true }
            title="Title"
            nextSteps={[]}
        >
        </ProcessStep>
    )).toThrowError();
});
