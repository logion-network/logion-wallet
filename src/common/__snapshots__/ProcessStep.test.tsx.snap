// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Rendering inactive step 1`] = `null`;

exports[`Rendering non-closable active step, may not proceed 1`] = `
<Dialog
  actions={
    Array [
      Object {
        "buttonText": "Proceed",
        "buttonVariant": "primary",
        "callback": [MockFunction],
        "choices": undefined,
        "disabled": true,
        "id": "proceed",
      },
    ]
  }
  show={true}
  size="lg"
>
  <React.Fragment>
    <h2>
      Title
    </h2>
    <p>
      Content
    </p>
  </React.Fragment>
</Dialog>
`;

exports[`Rendering non-closable active step, may proceed 1`] = `
<Dialog
  actions={
    Array [
      Object {
        "buttonText": "Proceed",
        "buttonVariant": "primary",
        "callback": [MockFunction],
        "choices": undefined,
        "disabled": false,
        "id": "proceed",
      },
    ]
  }
  show={true}
  size="lg"
>
  <React.Fragment>
    <h2>
      Title
    </h2>
    <p>
      Content
    </p>
  </React.Fragment>
</Dialog>
`;
